sympy/sets/sets.py

```diff
@@ -201,8 +203,8 @@ def _complement(self, other):
             # XXX: this doesn't work if the dimensions of the sets isn't same.
             # A - B is essentially same as A if B has a different
             # dimensionality than A
-            switch_sets = ProductSet(FiniteSet(o, o - s) for s, o in
-                                     zip(self.sets, other.sets))
+            switch_sets = ProductSet(*(FiniteSet(o, o - s) for s, o in
+                                     zip(self.sets, other.sets)))
```

Comment:
There are still some issues with like
```
A = FiniteSet(*symbols('a:c'))
B = FiniteSet(*symbols('d:f'))
Complement(ProductSet(A, A), B)
```
giving error.
But I would rather think that it's the flaw of the original code.

Reply:
The flaw is that it doesn't check that both sets are ProductSets. I've fixed this in the last commits.